[Data] [Docs] Update data stats docs (#44164)
ray-project/ray: doc/source/data/monitoring-your-workload.rst

summary across many different blocks, so some stats show the min, max, mean, and sum of the stat aggregated over all the blocks.
The following are descriptions of the various stats included at the operator level:

* **Remote wall time**: The wall time is the start to finish time for an operator. It includes the time where the operator
  isn't processing data, sleeping, waiting for I/O, etc.
* **Remote CPU time**: The CPU time is the process time for an operator which excludes time slept. This includes both
  user and system CPU time.
* **UDF time**: The UDF time is time spent in functions defined by the user. This includes functions passed into Ray
  Data methods, including :meth:`~ray.data.Dataset.map`, :meth:`~ray.data.Dataset.map_batches`, :meth:`~ray.data.Dataset.filter`,
  etc. You can use this to track the time spent in functions you define and how much time optimizing those functions could save.
* **Memory usage**: The output displays memory usage per block in MiB.
* **Output stats**: The output includes stats on the number of rows output and size of output in bytes per block. The number of

>   output rows per task are also included. All of this together gives you insight into how much data is being output at a per

Reviewer: ```suggestion
  output rows per task are also included. All of this together gives you insight into how much data Ray Data is outputting at a per
```

  block and per task level.
* **Task Stats**: The output shows the scheduling of tasks to nodes, which allows you to see if you are utilizing all of your nodes
  as expected.
* **Throughput**: The summary calculates the throughput for the operator, and for a point of comparison, it also computes an estimate of
  the throughput of the same task on a single node. This estimate assumes the total time of the work remains the same, but with no
  concurrency. The overall summary also calculates the throughput at the dataset level, including a single node estimate.

Iterator stats
~~~~~~~~~~~~~~

> If you iterate over the data, iteration stats are also generated. Even if you aren't directly iterating over the data, there

Reviewer: ```suggestion
If you iterate over the data, Ray Data also generates iteration stats. Even if you aren't directly iterating over the data, you
```

> might be iteration stats included (for example, if you call :meth:`~ray.data.Dataset.take_all`). Some of the stats included at the

Reviewer: ```suggestion
might get iteration stats, for example, if you call :meth:`~ray.data.Dataset.take_all`). Some of the stats that Ray Data includes at the
```

iterator level are:

> * **Iterator initialization**: The time spent initializing the iterator. This is internal to Ray Data.

Reviewer: ```suggestion
* **Iterator initialization**: The time Ray Data spent initializing the iterator. This time is internal to Ray Data.
```

> * **Time user thread is blocked**: The time spent producing data in the iterator. This is often the primary execution of a

Reviewer: ```suggestion
* **Time user thread is blocked**: The time Ray Data spent producing data in the iterator. This time is often the primary execution of a
```

  dataset if you haven't previously materialized it.
* **Time in user thread**: The time spent in the user thread that's iterating over the dataset outside of the Ray Data code.

>   If this time is high consider, optimizing the body of the loop that's iterating over the dataset.

Reviewer: ```suggestion
  If this time is high, consider optimizing the body of the loop that's iterating over the dataset.
```

> * **Batch iteration stats**: Stats are also included about the prefetching of batches. These times are internal to Ray

Reviewer: ```suggestion
* **Batch iteration stats**: Ray Data also includes stats about the prefetching of batches. These times are internal to Ray
```

>   Data code, but you can further optimize this by tuning the prefetching process.

Reviewer: ```suggestion
  Data code, but you can further optimize this time by tuning the prefetching process.
```

Verbose stats
~~~~~~~~~~~~~~

> By default, only the most important high-level stats are logged. If you would like to enable verbose stats outputs, include

Reviewer: ```suggestion
By default, Ray Data only logs the most important high-level stats. To enable verbose stats outputs, include
```

the following snippet in your Ray Data code:

.. testcode::

   from ray.data import DataContext

   context = DataContext.get_current()
   context.verbose_stats_logs = True

> By enabling verbosity there are a few more outputs added:

Reviewer: ```suggestion
By enabling verbosity Ray Data adds a few more outputs:
```

* **Extra metrics**: Operators, executors, etc. can add to this dictionary of various metrics. There is
  some duplication of stats between the default output and this dictionary, but for advanced users provides more insight into
  the dataset's execution.

> * **Runtime metrics**: High level breakdown of the runtime of the dataset execution. This is a per operator summary of the

Reviewer: ```suggestion
* **Runtime metrics**: These metrics are a high-level breakdown of the runtime of the dataset execution. This stat is a per operator summary of the
```

>   time each operator took to complete and the fraction of the total execution time that operator took to complete. As there

Reviewer: ```suggestion
  time each operator took to complete and the fraction of the total execution time that the operator took to complete. As there
```

  is not processing data, sleeping, waiting for I/O, etc.
* **Remote cpu time**: The cpu time is the process time for an operator which will exclude time slept. This will include 
  both user and system CPU time.
* **UDF time**: The UDF time is time spent in functions defined by the user. This will include functions passed into Ray 
  Data methods including :meth:`~ray.data.Dataset.map`, :meth:`~ray.data.Dataset.map_batches`, :meth:`~ray.data.Dataset.filter`,
  etc. You can use this to track how much time is spent in functions you define and how much time could be gained by optimizing 
  those functions.
* **Memory usage**: The memory usage per block is shown in MiB. 
* **Output stats**: The output includes stats on the number of rows output and size of output in bytes per block. The number of
  output rows per task are also included. All of this together will give you insight into how much data is being output at a per
  block and per task level.
* **Task Stats**: The scheduling of tasks to nodes is shown which will allow you to see if you are utilizing all of your nodes
  as expected.

> * **Throughput**: Throughput for the operator is calculated and for a point of comparison an estimate of the throughput of the 
>   same task on a single node is computed. This estimate assumes the total time of the work remains the same, but there is no 
>   concurrency. Throughput is also calculated at the dataset level including a single node estimate.

Reviewer 2: ```suggestion
* **Throughput**: Throughput for the operator is calculated, and as a point of comparison, an estimate of the throughput of the 
  same task on a single node is computed. This estimate assumes the total time of the work remains the same, but with no 
  concurrency. Throughput is also calculated at the dataset level, including a single node estimate.
```

Iterator stats
~~~~~~~~~~~~~~

> If the data is iterated over, iteration stats are also generated. Even if you are not directly iterating over the data, there might be iteration stats included e.g. if you 
> call :meth:`~ray.data.Dataset.take_all`. Some of the stats included at the iterator level are:

Reviewer 2: (avoid using abbreviations for Vale styling)
```suggestion
If the data is iterated over, iteration stats are also generated. Even if you are not directly iterating over the data, there might be iteration stats included (for example, if you 
call :meth:`~ray.data.Dataset.take_all`). Some of the stats included at the iterator level are:
```

> * **Iterator initialization**: The time spent initializing the iterator to be used. This is internal to Ray Data.

Reviewer 2: ```suggestion
* **Iterator initialization**: The time spent initializing the iterator. This is internal to Ray Data.
```

> * **Time user thread is blocked**: The time spent producing data in the iterator. Often this will be the primary execution of a

Reviewer 2: ```suggestion
* **Time user thread is blocked**: The time spent producing data in the iterator. This is often the primary execution of a
```

  dataset if it has not previously been materialized. 
* **Time in user thread**: The time spent in the user thread that is iterating over the dataset outside of the Ray Data code.

>   If this time is high consider optimizing the body of the loop that is iterating over the dataset.

Reviewer 2: ```suggestion
  If this time is high, consider optimizing the body of the loop that is iterating over the dataset.
```

> * **Batch iteration stats**: Stats are also included about the prefetching of batches. These times should be internal to Ray
>   Data code, but by tuning how prefetching is performed this might be further optimized.

Reviewer 2: ```suggestion
* **Batch iteration stats**: Stats are also included about the prefetching of batches. These times are internal to Ray
  Data code, but by tuning how prefetching is performed, this can be further optimized.
```

Example stats
~~~~~~~~~~~~~

> For a concrete example, below is a stats output from :doc:`Image Classification Batch Inference with PyTorch ResNet18 </data/examples/pytorch_resnet_batch_prediction>`:

Reviewer 2: ```suggestion
As a concrete example, below is a stats output from :doc:`Image Classification Batch Inference with PyTorch ResNet18 </data/examples/pytorch_resnet_batch_prediction>`:
```

  concurrency. Throughput is also calculated at the dataset level including, a single node estimate.

Iterator stats
~~~~~~~~~~~~~~
If the data is iterated over, iteration stats are also generated. Even if you aren't directly iterating over the data, there
might be iteration stats included (for example, if you call :meth:`~ray.data.Dataset.take_all`). Some of the stats included at the
iterator level are:

* **Iterator initialization**: The time spent initializing the iterator. This is internal to Ray Data.
* **Time user thread is blocked**: The time spent producing data in the iterator. This is often the primary execution of a
  dataset if it hasn't previously been materialized.
* **Time in user thread**: The time spent in the user thread that's iterating over the dataset outside of the Ray Data code.
  If this time is high consider, optimizing the body of the loop that's iterating over the dataset.
* **Batch iteration stats**: Stats are also included about the prefetching of batches. These times are internal to Ray
  Data code, but by tuning how prefetching is performed, this can be further optimized.

Verbose stats
~~~~~~~~~~~~~~

> By default, the verbosity of the stats output isn't enabled. If you would like to enable verbose stats outputs, include

Reviewer 2: ```suggestion
By default, only the most important high-level stats are logged. If you would like to enable verbose stats outputs, include
```

the following snippet in your Ray Data code:

.. code-block:: python

   from ray.data import DataContext

   context = DataContext.get_current()
   context.verbose_stats_logs = True

By enabling verbosity there are a few more outputs added:

* **Extra metrics**: A dictionary of various metrics which is added to by operators, executors, etc. There is

>   some duplication of stats between the default output and this dictionary, but for advanced users provides more insight into

Reviewer 2: ```suggestion
  some duplication of stats between the default output and this dictionary, but for advanced users, this provides more insight into
```

  the dataset's execution.
* **Runtime metrics**: High level breakdown of the runtime of the dataset execution. This is a per operator summary of the
  time each operator took to complete and the fraction of the total execution time that operator took to complete. As there

>   are potentially multiple concurrent operators, these percentages don't necessarily sum to 100%. Instead this is meant
>   to show how long running each of the operators is in the context of the full dataset execution.

Reviewer 2: ```suggestion
  are potentially multiple concurrent operators, these percentages don't necessarily sum to 100%. Instead,
  they show how long running each of the operators is in the context of the full dataset execution.
```
